### PR TITLE
add matmul 2 code transformation to psyad (closes 1559)

### PR DESCRIPTION
**Pull request by rupertford**

### CODE REVIEW

**Comment by rupertford** [comment]
This is ready for first review by @arporter subject to CI

File: doc/psyad/user_guide/implementation.rst [hunk 1]
```diff
@@ -374,17 +374,20 @@ it must be transformed to its associated adjoint form.
...
-
-.. note:: At the moment all ``dot_product`` instrinsics are transformed
-	  irrespective of whether they act on active variables or not.
+The only supported intrinsics at this time are ``dot_product`` and
+``matmul``.
+
+If a ``dot_product`` or ``matmul`` intrinsic is found in the
+tangent-linear code it is first transformed into equivalent inline
+code before the code is transformed to its adjoint form. The PSyIR
+``DotProduct2CodeTrans`` or ``Matmul2CodeTrans`` transformation
```
Comment (arporter) [inline doc/psyad/user_guide/implementation.rst:383]: "...transformations are used by PSyAD to perform these transformations."

File: src/psyclone/psyad/transformations/preprocess.py [hunk 2]
```diff
@@ -54,8 +55,14 @@ def preprocess_trans(kernel_psyir):
...
     '''
     dot_product_trans = DotProduct2CodeTrans()
+    matmul_trans = Matmul2CodeTrans()
 
     for oper in kernel_psyir.walk(BinaryOperation):
         if oper.operator == BinaryOperation.Operator.DOT_PRODUCT:
             # Apply DOT_PRODUCT transformation
             dot_product_trans.apply(oper)
+
+    for oper in kernel_psyir.walk(BinaryOperation):
```
Comment (arporter) [inline src/psyclone/psyad/transformations/preprocess.py:65]: You could do away with this separate loop (and thus its associated `walk`) and just have the `if` as a second child of the loop at line 60.

**Review by arporter — CHANGES_REQUESTED** [review]
Good job. Just a few minor things to fix/tidy up.
Documentation has been updated and builds OK.
Tests and examples are all OK with compilation.
Associated test file gives 100% coverage of modified source file.
Issue number is not mentioned anywhere.

File: doc/psyad/user_guide/implementation.rst (hunk 1 above)
Comment (rupertford) [reply doc/psyad/user_guide/implementation.rst:383]: Better, thanks.

File: src/psyclone/psyad/transformations/preprocess.py (hunk 2 above)
Comment (rupertford) [reply src/psyclone/psyad/transformations/preprocess.py]: As discussed privately I was concerned about the transformations affecting each other. However this is not the case, so I have done as you suggest. When we add support nested intrinsics (see #1561 and #1562 for example) we might need to revisit - and perhaps even have an outer `while intrinsics` loop. However, the tests associated with implementing solutions to those issues should pick up any problems.

**Review by arporter — APPROVED** [review]
All requested changes have been made.
Examples and tests OK with compilation.
All docs build OK modulo one minor error in kernel_tools which I've fixed.

**Pull request merged**